src/index.ts

```diff
@@ -490,123 +515,255 @@ export class Upload extends Pumpify {
+   */
+  private async *upstreamIterator(limit: number, multiChunkMode: boolean) {
+    let completeChunk = Buffer.alloc(0);
+
+    // read from upstream chunk buffer
+    while (limit && (await this.waitForNextChunk())) {
+      // read until end or limit has been reached
+      const chunk = this.pullFromChunkBuffer(limit);
+
+      limit -= chunk.byteLength;
+      if (multiChunkMode) {
+        // return 1 chunk at the end of iteration
+        completeChunk = Buffer.concat([completeChunk, chunk]);
+      } else {
+        // return many chunks throughout iteration
+        yield {
+          chunk,
+          encoding: this.chunkBufferEncoding,
+        };
       }
-    };
+    }
 
-    // This should be 'once' as `startUploading` can be called again
-    // for multi chunk uploads.
-    this.once('response', responseHandler);
+    if (multiChunkMode) {
+      yield {
+        chunk: completeChunk,
+        encoding: this.chunkBufferEncoding,
+      };
+    }
+  }
+
+  async startUploading() {
+    const multiChunkMode = !!this.chunkSize;
+    let expectedUploadSize: number | undefined = undefined;
+
+    if (!this.offset) {
+      this.offset = 0;
+    }
+
+    if (this.chunkSize && typeof this.contentLength === 'number') {
+      const remainingBytesToUpload = this.contentLength - this.numBytesWritten;
+
+      // Uploads should be no more than the `chunkSize`
+      expectedUploadSize = Math.min(this.chunkSize, remainingBytesToUpload);
+    }
 
-    this.setPipeline(
-      bufferStream,
-      chunkSizingStream,
-      offsetStream,
-      delayStream
+    // A queue for the upstream data
+    const upstreamQueue = this.upstreamIterator(
+      expectedUploadSize || Infinity,
+      multiChunkMode // multi-chunk mode should return 1 chunk per request
     );
 
-    this.pipe(requestStreamEmbeddedStream);
+    // The primary read stream for this request. This stream retrieves no more
+    // than the exact requested amount from upstream.
+    const leadStream = new Readable({
+      read: async () => {
+        const result = await upstreamQueue.next();
+
+        if (result.value) {
+          if (this.chunkSize) {
+            // Cache the data in case the server misses some bytes in the request.
+            // We should not assume that the server received all bytes sent in the request.
+            // - https://cloud.google.com/storage/docs/performing-resumable-uploads#chunked-upload
+            this.multiChunkUploadChunk = result.value.chunk;
+          }
+
+          if (this.numBytesWritten === 0) {
+            if (!this.checkFirstChunkOfUpload(result.value.chunk)) {
+              this.unshiftChunkBuffer(result.value.chunk);
+              this.restart();
+              leadStream.push(null);
+              return;
+            }
+          }
+
+          leadStream.push(result.value.chunk, result.value.encoding);
+        }
+
+        if (result.done) {
+          leadStream.push(null);
```

Comment:
I didn't look closely but just want to verify that using `null` as the argument here won't cause problems?

Reply:
`null` tells signals to consumers 'no more data will be available from this stream': https://nodejs.org/api/stream.html#readablepushchunk-encoding